APILambda/validation.ts

```diff
@@ -38,62 +38,108 @@ interface ValidationSchemas {
+        params: pathParamsSchema ?? z.object({}).strict()
+      })
 
-      if (pathParamsSchema) {
-        req.params = pathParamsSchema.parse(req.params)
-      } else if (Object.keys(req.params).length !== 0) {
-        throw new Error("Unexpected path parameters provided")
-      }
+      await validationSchema.parseAsync({
+        body: req.body,
+        query: req.query,
+        params: req.params
+      })
 
       next()
     } catch (error) {
-      console.log("failed to validate request ", error)
-      return res.status(400).json({ error: "invalid-request" })
+      if (error instanceof z.ZodError) {
+        console.log("failed to validate request ", error)
+        return res.status(400).json({ error: "invalid-request" })
+      } else {
+        console.error("Error in validation middleware: ", error)
+        return res.status(500).json({ error: "internal-server-error" })
+      }
     }
   }
 
+type InferRequestSchemaType<T> = T extends ZodSchema<any> ? z.infer<T> : never;
+
+type ValidatedRequestHandler<S extends ValidationSchemas> = (
+  req: Omit<Request, "body" | "query" | "params"> & {
+    body: InferRequestSchemaType<S["bodySchema"]>;
+    query: InferRequestSchemaType<S["querySchema"]>;
+    params: InferRequestSchemaType<S["pathParamsSchema"]>;
+  },
+  res: Response,
+  next: NextFunction
+) => Promise<Response>;
+
+/**
+ * Wrapper around the Express Router which facilitates runtime validation and
+ * TypeScript type inference for request handlers based on Zod schemas.
+ *
+ * Example Usage:
+ * ```typescript
+ * import { z } from 'zod';
+ * import { ValidatedRouter } from './validated-router';  // Assume the file is named validated-router.ts
+ *
+ * // Define your schemas
+ * const bodySchema = z.object({
+ *   name: z.string(),
+ *   age: z.number().positive(),
+ * });
+ *
+ * const querySchema = z.object({
+ *   sortBy: z.string().optional(),
+ * });
+ *
+ * // Create an instance of ValidatedRouter
+ * const router = new ValidatedRouter();
+ *
+ * // Define your route
+ * router.post('/create-user', { bodySchema, querySchema }, (req, res) => {
+ *   // At this point, req.body and req.query are correctly typed
+ *   // and have been validated to match the schemas.
+ *   // ...
+ * });
+ *
+ * // To use this router in your express application:
+ * app.use('/api', router.asRouter());
+ * ```
+ *
+ * @class ValidatedRouter
+ */
 export class ValidatedRouter {
   private router: Router
 
   constructor () {
     this.router = express.Router()
   }
 
-  get (path: string, schema: Pick<ValidationSchemas, "querySchema" | "pathParamsSchema">, ...handlers: RequestHandler[]): this {
-    this.router.get(path, validateRequest(schema), ...handlers)
+  get<Schema extends ValidationSchemas> (path: string, schema: Pick<Schema, "querySchema" | "pathParamsSchema">, ...handlers: ValidatedRequestHandler<Schema>[]): this {
+    this.router.get(path, validateRequest(schema), ...handlers as any)
```

Comment:
Since the ValidatedRequest handler and express Request handler have different types, I assigned an “any” when passing it down so ts doesn’t complain. May find a better solution later.